mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -587,6 +581,116 @@ static Value reconfigureLAGeneric(PatternRewriter &b,
   return laOut;
 }
 
+static LogicalResult findGlobalStore(linalg::GenericOp laGeneric,
+                                     OpOperand *&inputLeadingToGlobalStore,
+                                     GlobalStoreOp &gemmStoreOp) {
+  for (auto pair : llvm::zip(laGeneric.getInputOperands(),
+                             laGeneric.getIndexingMapsArray())) {
+    AffineMap inpIdxMap = std::get<1>(pair);
+    OpOperand *input = std::get<0>(pair);
+    GlobalStoreOp maybeStore = traceToGlobalStore(input->get());
+    if (maybeStore) {
+      if (gemmStoreOp) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "Multiple generic inputs come from writeback\n");
+        return failure();
+      }
+
+      auto aliasingOutputMap =
+          findAliasingOutputIdxMap(laGeneric, input->get());
+      AffineMap invertOutIdxMap = inversePermutation(aliasingOutputMap);
+      auto outToInMap = inpIdxMap.compose(invertOutIdxMap);
+      SmallVector<unsigned> permutedDims;
+      // This is not to allow broadcasting but due to canonical linalg
+      // form if the unit dims can carry affine expr to be zero in the
+      // translation, hence there is a following check.
+      if (!outToInMap.isMinorIdentityWithBroadcasting(&permutedDims)) {
+        LLVM_DEBUG(llvm::dbgs() << outToInMap << "\n");
+        LLVM_DEBUG(
+            llvm::dbgs()
+            << "The store is not even a minor identity with broadcasting.\n");
+        return failure();
+      }
+      auto inpShape = input->get().getType().cast<ShapedType>().getShape();
+      LLVM_DEBUG(llvm::dbgs() << "outToInMap = " << outToInMap << "\n");
+      LLVM_DEBUG(llvm::dbgs()
+                 << "inp shape = " << input->get().getType().cast<ShapedType>()
+                 << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "permutedDims = ");
+      LLVM_DEBUG(llvm::interleaveComma(permutedDims, llvm::dbgs()));
+      LLVM_DEBUG(llvm::dbgs() << "\n");
+
+      for (auto bDim : permutedDims) {
+        if (inpShape[bDim] != 1) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "The store input cannot be a real broacast.\n");
+          return failure();
+        };
+      }
+      gemmStoreOp = maybeStore;
+      inputLeadingToGlobalStore = input;
+    }
+  }
+  if (!gemmStoreOp) {
+    LLVM_DEBUG(llvm::dbgs() << "No input is leading to a global store.\n");
+    return failure();
+  }
+  return success();
+}
+
+// This function will add a TransformingFor to translate original coordinates
+// of the gemmStore op to be transformed to match the aliasing shape of the
+// output. This is done by replacing the original gemmStoreOp with a new one
+// wrapped in a TransformingFor op.
+static void adjustStoreCoordsForDestArg(
```

Comment:
Note: the OOB check I had there can potentially break in the presence of the Unfold operator (which currently comes up in the transforms for filter tensors and so won't be an issue here but...)

Also, for our consideration, whether adding the ability to add new domains to existing loops is a cleaner way to do some of this.

Will elaborate tomorrow afternoon